File: android/app/src/main/java/xyz/juniverse/babylistener/firebase/DevicePairing.kt
package xyz.juniverse.babylistener.firebase

import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.*
import xyz.juniverse.babylistener.etc.Pref
import xyz.juniverse.babylistener.etc.console
import java.util.*

/**
 * Created by juniverse on 03/01/2018.
 *
 * todo
 * 'code' : {
 *      'reqId' : '$uid'
 *      'time' : 123456
 *      'ackId' : '$uid'
 * }
 */
class DevicePairing {
    companion object {
        private var _instance: DevicePairing? = null
        fun getInstance(): DevicePairing {
            if (_instance == null)
                _instance = DevicePairing()
            return _instance!!
        }

        val SUCCESS = 0
        val UNKNOWN_CODE = -1
        val TIMEOUT = -2
    }

    private data class PairingData(val reqId: String = "", val time: Long = 0, var ackId: String? = null)


//    private val maxWaitTime: Long = 3 * 60 * 1000      // 3 minutes
    private val maxWaitTime: Long = 30 * 1000      // 30 seconds
    private val pairDbRef
        get() = FirebaseDatabase.getInstance().getReference(DB.Table.pair)



    fun registerTest() {
        val currentUser = FirebaseAuth.getInstance().currentUser ?: return

        console.d("I am ", currentUser.uid)
        pairDbRef.addListenerForSingleValueEvent(object: ValueEventListener{
            override fun onCancelled(error: DatabaseError?) {
                console.d("value event onCancelled", error)
            }
            override fun onDataChange(snapshot: DataSnapshot) {
                console.d("value event onDataChange", snapshot)
                val key = "999999"
                if (!snapshot.children.any { it.key == key }) {
                    console.d("adding pairing request..")
                    pairDbRef.child(key).setValue(PairingData(currentUser.uid, Date().time))
                    pairDbRef.removeEventListener(this)
                } else {
                    console.d("key already in... cannot write")
                }
            }
        })
    }




    fun registerPairRequest(onResult: (String?) -> Unit) {
        val currentUser = FirebaseAuth.getInstance().currentUser ?: return onResult(null)

        console.d("making request")

        pairDbRef.addListenerForSingleValueEvent(object: ValueEventListener{
            override fun onCancelled(error: DatabaseError?) {
                console.d("onCancelled", error)
            }
            override fun onDataChange(snapshot: DataSnapshot) {
                console.d("value event onDataChange", snapshot)
                var code: String
                do {
                    code = (Random().nextInt(899999) + 100000).toString()
                } while (snapshot.children.any { it.key == code })

                console.d("adding pairing request..")
                pairDbRef.child(code).setValue(PairingData(currentUser.uid, Date().time))

                onResult(code)
            }
        })
    }


    fun cancelWaitingForAck(code: String?) {
        if (code == null) return

        console.d("cancelWaitingForAck")
        cancelTimer?.cancel()
        pairDbRef.child(code).removeEventListener(acknowledgeListener)
        pairDbRef.child(code).removeValue()
        acknowledgeListener = null
    }

    // wait for that code to change. if false, timeout
    private var acknowledgeListener: ValueEventListener? = null
    private var cancelTimer: Timer? = null
    fun waitForAcknowledge(code: String, onResult: (Boolean) -> Unit) {
        if (acknowledgeListener != null) return

        val codeDatRef = pairDbRef.child(code)
        acknowledgeListener = object : ValueEventListener {
            override fun onCancelled(databaseError: DatabaseError) {
                // just in case
                console.d("onCancelled", databaseError)
                codeDatRef.removeEventListener(this)
            }

            override fun onDataChange(dataSnapshot: DataSnapshot) {
                console.d("onDataChange", dataSnapshot)

                val pairingData = dataSnapshot.getValue(PairingData::class.java)
                if (pairingData == null) {
                    // this is when it's canceled.. though this probably won't happen
                    cancelWaitingForAck(code)
                    return
                }

                val ackId = pairingData.ackId ?: return

                console.d("answered!!! key?", dataSnapshot.key)
                console.d("pairingData?", pairingData)

                // save
                savePartnerId(ackId)

                // remove listener
                cancelWaitingForAck(code)

                onResult(true)
            }
        }
        codeDatRef.addValueEventListener(acknowledgeListener)

        cancelTimer = Timer()
        cancelTimer?.schedule(object: TimerTask() {
            override fun run() {
                cancelWaitingForAck(code)
                onResult(false)
            }
        }, maxWaitTime)
    }






    // ============ for ackId ==============
    fun acknowledgePairing(code: String, complete: (Int) -> Unit): Boolean {
        val currentUser = FirebaseAuth.getInstance().currentUser ?: return false

        pairDbRef.child(code).addListenerForSingleValueEvent(object: ValueEventListener {
            override fun onCancelled(error: DatabaseError?) {
                console.e("acknowledge error?", error)
            }
            override fun onDataChange(snapshot: DataSnapshot) {
                val pairingData = snapshot.getValue(PairingData::class.java)
                if (pairingData == null) {
                    // todo possibly removed by timeout
                    complete(UNKNOWN_CODE)
                    return
                }

                if (Date().time - pairingData.time > maxWaitTime) {
                    complete(TIMEOUT)
                    return
                }

                pairingData.ackId = currentUser.uid
                pairDbRef.child(code).setValue(pairingData)

                complete(SUCCESS)

                savePartnerId(pairingData.reqId)
            }
        })

        return true
    }


    fun unpairDevice() {
        // todo notify partner to unpair
    }


    private fun savePartnerId(partnerId: String) {
        // todo after math
        console.d("I am", FirebaseAuth.getInstance().currentUser?.uid, "and my partner is", partnerId)
//        Pref.putString(Pref.partnerId, partnerId)
    }

    fun checkPartnerId(result: (Boolean) -> Unit) {
        val partnerId = Pref.getString(Pref.partnerId) ?: return result(false)

        // todo need to send some msg to acknowledge
        result(true)
    }
}

File: ios/BabyListener/firebase/DevicePairing.swift
//
//  DevicePairing.swift
//  BabyListener
//
//  Created by 김준성 on 05/01/2018.
//  Copyright © 2018 김준성. All rights reserved.
//

import Foundation
import Firebase

class DevicePairing {
    static let SUCCESS = 0
    static let UNKNOWN_CODE = -1
    static let TIMEOUT = -2
    
    let PAIR_TABLE_NAME = "pairing"
    let maxWaitingTime = 3 * 60 * 1000
    
    func registerPairRequest(onResult: @escaping (_ result: String?) -> Void) {
        guard let userId = Auth.auth().currentUser?.uid else {
            print("not logged in")
            onResult(nil)
            return
        }
        
        print("making request")
        let pairDbRef = Database.database().reference(withPath: PAIR_TABLE_NAME)
        pairDbRef.observeSingleEvent(of: DataEventType.value, with: {(snapshot) in
            var code: String
            var found: Bool
            repeat {
                found = false
                code = String(arc4random_uniform(900000) + 100000)
                print("checking code \(code)")
                let enumerator = snapshot.children
                while let pairReq = enumerator.nextObject() as? DataSnapshot {
                    if pairReq.key == code {
                        found = true
                        break
                    }
                }
            } while (found)
            
            let pairData = ["reqId": userId,
                            "time": UInt(NSDate().timeIntervalSince1970 * 1000)] as [String : Any]
//            let pairData = PairingData(reqId: userId, time: UInt(CACurrentMediaTime() * 1000), ackId: nil)
            print("adding pairing request \(pairData)")
            pairDbRef.child(code).setValue(pairData)
           
            onResult(code)
        })
    }
    
    var acknowledgeObserverHandle: UInt = 0
    func waitForAcknowledge(code: String, onResult: @escaping (_ result: Bool) -> Void) {
        if acknowledgeObserverHandle != 0 { return }
        
        let codeDataRef = Database.database().reference(withPath: PAIR_TABLE_NAME).child(code)
            acknowledgeObserverHandle = codeDataRef.observe(DataEventType.value, with: {(snapshot) in
//                print("compiling???")
                guard let pairingData = snapshot.value as? [String : Any] else {
                    self.cancelWaitingForAck(code)
                    return
                }
                
                guard let ackId = pairingData["ackId"] as? String else {
                    return
                }
                
                print("answered!!! \(ackId)")
                self.savePartnerId(partnerId: ackId)
                
                self.cancelWaitingForAck(code)
                
                onResult(true)
                })
        
        print("acknowledgeObserverHandle? \(acknowledgeObserverHandle)")
        // todo timer
    }
    
    func cancelWaitingForAck(_ code: String) {
        // todo
     
        let codeDataRef = Database.database().reference(withPath: PAIR_TABLE_NAME).child(code)
        codeDataRef.removeObserver(withHandle: acknowledgeObserverHandle)
        codeDataRef.removeValue()
        acknowledgeObserverHandle = 0
    }
    
    func acknowledgePairing(code: String, onResult: @escaping (_ result: Int) -> Void) -> Bool {
        guard let userId = Auth.auth().currentUser?.uid else {
            print("not logged in")
            return false
        }
        
        let codeDataRef = Database.database().reference(withPath: PAIR_TABLE_NAME).child(code)
        codeDataRef.observeSingleEvent(of: DataEventType.value, with: {(snapshot) in
            guard let pairingData = snapshot.value as? [String : Any] else {
                onResult(DevicePairing.UNKNOWN_CODE)
                return
            }
            
            let now = UInt(NSDate().timeIntervalSince1970 * 1000)
            if now - (pairingData["time"] as! UInt) > self.maxWaitingTime {
                onResult(DevicePairing.TIMEOUT)
                return
            }
            
            let updateData = ["reqId": pairingData["reqId"] as! String,
                            "time": pairingData["time"] as! UInt,
                            "ackId": userId] as [String : Any]
            codeDataRef.updateChildValues(updateData)
            
            onResult(DevicePairing.SUCCESS)
            
            self.savePartnerId(partnerId: pairingData["reqId"] as! String)
        })
        return true
    }
    
    func unpairDevice() {
        // todo
    }
    
    func savePartnerId(partnerId: String) {
        // todo
        let userId = Auth.auth().currentUser?.uid
        print("I am \(userId) and my parnter is \(partnerId)")
    }
    
    func checkPartnerId(onResult: (_ result: Bool) -> Void) {
        // todo
    }
}
